### PR TITLE
[video_player] Optimize caption retrieval with binary search in VideoPlayerController

### PR DESCRIPTION
**Pull request by abdelaziz-mahdy**
Just found the todo in the code and it was from 3 years ago, so i wanted to implement it since it may help people using captions to have more optimized code

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

### CODE REVIEW

File: packages/video_player/video_player/lib/video_player.dart [hunk 1]
```diff
@@ -728,20 +728,36 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
...
-    for (final Caption caption in _closedCaptionFile!.captions) {
-      if (caption.start <= delayedPosition && caption.end >= delayedPosition) {
-        return caption;
+
+    final List<Caption> captions = _closedCaptionFile!.captions;
+
+    int left = 0;
+    int right = captions.length - 1;
+
+    while (left <= right) {
```
Comment (jmagman) [inline packages/video_player/video_player/lib/video_player.dart:744]: https://github.com/flutter/plugins/pull/2487#discussion_r376564502

> Not a big deal either way, but for what it's worth I think you could trivially import this. https://api.flutter.dev/flutter/package-collection_collection/binarySearch.html

> Yes, it wouldn't be too hard, although we would have to either sort the captions at initialization, or make it clear that we assume a duration sorted list of captions (neither of which we currently do).

Comment (abdelaziz-mahdy) [reply packages/video_player/video_player/lib/video_player.dart]: I implemented all the mentioned changes. However, to use the collection function, I had to convert to a list, so I cached it to avoid re-mapping on each check.

Comment (tarrinneal) [reply packages/video_player/video_player/lib/video_player.dart]: I think what @jmagman is suggesting is that there is a package should be used to do the binary search.

File: packages/video_player/video_player/CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,7 @@
 ## NEXT
```
Comment (tarrinneal) [inline packages/video_player/video_player/CHANGELOG.md:1]: This will need a version change to 2.9.3, as well as in the pubspec file

**Review by tarrinneal — CHANGES_REQUESTED** [review]

**Comment by tarrinneal** [comment]
@stuartmorgan does this change need a test? It shouldn't be changing any behaviors, I'm not sure how a test would work honestly.

**Comment by stuartmorgan** [comment]
> @stuartmorgan does this change need a test? It shouldn't be changing any behaviors, I'm not sure how a test would work honestly.

The fact that it's a more efficient search doesn't inherently need a test; that's essentially just a refactoring. However, this seems like a good opportunity to make sure we have reasonably robust coverage of caption correctness. E.g., is there a unit test that would fail if the initial sort were commented out?

**Comment by abdelaziz-mahdy** [comment]
> > @stuartmorgan does this change need a test? It shouldn't be changing any behaviors, I'm not sure how a test would work honestly.
> 
> The fact that it's a more efficient search doesn't inherently need a test; that's essentially just a refactoring. However, this seems like a good opportunity to make sure we have reasonably robust coverage of caption correctness. E.g., is there a unit test that would fail if the initial sort were commented out?

I can add that , so in general more caption edge cases

**Comment by abdelaziz-mahdy** [comment]
tests implemented and improved the logic for the binary search since it was incorrect in some cases

for this test `test('captions are sorted correctly on initialization'`
this test can be ignored and if thats the case we can remove the new public member to the class (my main issue is that i was not able to expose a private member only for testing) if someone can point me to the right direction that will help a lot

File: packages/video_player/video_player/lib/video_player.dart [hunk 3]
```diff
@@ -378,6 +379,10 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
 
   Future<ClosedCaptionFile>? _closedCaptionFileFuture;
   ClosedCaptionFile? _closedCaptionFile;
+  List<Caption>? _sortedCaptions;
+
+  /// The sorted list of captions from the closed caption file.
+  List<Caption>? get sortedCaptions => _sortedCaptions;
```
Comment (stuartmorgan) [inline packages/video_player/video_player/lib/video_player.dart:385]: Why is new public API needed? I don't see any implementation code using this.

Comment (abdelaziz-mahdy) [reply packages/video_player/video_player/lib/video_player.dart:385]: To be able to access that list in testing, as I mentioned in the last comment I couldn't make it private and available for testing, 

If the test is not needed I can remove it with the test

Comment (abdelaziz-mahdy) [reply packages/video_player/video_player/lib/video_player.dart]: i removed it and the test.

Comment (stuartmorgan) [reply packages/video_player/video_player/lib/video_player.dart]: If something needs to be visible only for testing, it should be annotated `@visibleForTesting`.

It's not clear to me why even that would be necessary though; API to get the current caption is already public, and I would expect that it should be straightforward to construct a case where getting the caption for some time would be incorrect unless the captions are sorted internally.

Comment (abdelaziz-mahdy) [reply packages/video_player/video_player/lib/video_player.dart]:
```
abstract class ClosedCaptionFile {
  /// The full list of captions from a given file.
  ///
  /// The [captions] will be in the order that they appear in the given file.
  List<Caption> get captions;
}

```

its only a getter so calling sort on it doesnt sort the original list, this is why i copied the list and sorted it and the public method was for the testing part only, if the test is not needed the public method is not needed

but `_sortedCaptions` is needed

Comment (stuartmorgan) [reply packages/video_player/video_player/lib/video_player.dart]: > this is why i copied the list and sorted it
> [...]
> but `_sortedCaptions` is needed

I'm not sure what part of my comments led you to believe I don't understand `_sortedCaptions`. My comment is about the new public API.

> the public method was for the testing part only, if the test is not needed the public method is not needed

I still don't understand why you are presenting a binary choice between adding new public API and having testing. Whether or not the correct caption is fetched for a given time is already an observable property with existing public API. You should be able to construct a case where, with binary search, *the wrong caption is reported* without the internal sorting step, and thus validate that it is being handled correctly without directly inspecting internal state via new public accessors.

Testing internal implementation details is an anti-pattern, and should only be done when the important properties cannot be tested via the actual client APIs. I have not seen any argument for why that would not be the case here. What scenario are you unable to test without adding a new public API?

Comment (abdelaziz-mahdy) [reply packages/video_player/video_player/lib/video_player.dart]: i didnt mean all the testing of the testing is ignored, just this test `captions are sorted correctly on initialization` since it requires the expose of the public interface so i can check if they are sorted correctly 

`List<Caption>? get sortedCaptions` without it i cant get the sorted list and without the sorted list i cant make that test

but i did add test cases for not sorted captions and tested retrieving them which works correctly and if the sorting is removed they wont pass

so i think those are enough, and no need to introduce the public interface since all of the cases are covered as you mentioned

and i did remove the public method.

**Comment by stuartmorgan** [comment]
From triage: what's the status of this PR? Is it ready for re-review?

**Comment by abdelaziz-mahdy** [comment]
> From triage: what's the status of this PR? Is it ready for re-review?

Yes it's ready for a re-review

File: packages/video_player/video_player/pubspec.yaml [hunk 4]
```diff
@@ -25,11 +25,13 @@ dependencies:
   flutter:
     sdk: flutter
   html: ^0.15.0
+  collection: ^1.19.0
```
Comment (tarrinneal) [inline packages/video_player/video_player/pubspec.yaml:28]: You need to run the analyze tool as well to see what's failing - currently it's this line. These deps should be in alphabetical order.

**Review by tarrinneal — CHANGES_REQUESTED** [review]
Changelog doesn't match our guidelines. And a tip with the failing tests.

Seems good to me besides.

File: packages/video_player/video_player/pubspec.yaml (hunk 4 above)
Comment (tarrinneal) [reply packages/video_player/video_player/pubspec.yaml:28]: https://github.com/flutter/plugins/blob/master_archive/script/tool/README.md#commands

Comment (abdelaziz-mahdy) [reply packages/video_player/video_player/pubspec.yaml:28]: Thank you for the link, I sorted them.

File: packages/video_player/video_player/pubspec.yaml [hunk 5]
```diff
@@ -30,6 +31,7 @@ dependencies:
   video_player_platform_interface: ^6.2.0
   video_player_web: ^2.1.0
 
+
```
Comment (tarrinneal) [inline packages/video_player/video_player/pubspec.yaml:34]: This line needs to be removed too

Comment (abdelaziz-mahdy) [reply packages/video_player/video_player/pubspec.yaml]: done.

**Comment by tarrinneal** [comment]
Now I don't know why that test is failing. @stuartmorgan have you seen this type of failure with the analyze legacy tests before?

**Comment by stuartmorgan** [comment]
> Now I don't know why that test is failing. @stuartmorgan have you seen this type of failure with the analyze legacy tests before?

Yes. The output explains the problem and what needs to be changed to fix it, so I'm not following what the issue is.

File: packages/video_player/video_player/pubspec.yaml [hunk 6]
```diff
@@ -22,6 +22,7 @@ flutter:
         default_package: video_player_web
 
 dependencies:
+  collection: ^1.19.0
```
Comment (stuartmorgan) [inline packages/video_player/video_player/pubspec.yaml:25]: This constraint needs to be adjusted as described in the CI failure output.

Comment (tarrinneal) [reply packages/video_player/video_player/pubspec.yaml:25]: I didn't realize the fix was in the error. Sorry

Comment (tarrinneal) [reply packages/video_player/video_player/pubspec.yaml:25]: Needs to be version 1.18.0

Comment (abdelaziz-mahdy) [reply packages/video_player/video_player/pubspec.yaml]: Thank you very much! I didn’t find the error logs initially, but after you mentioned them, I looked further and found them.

**Comment by tarrinneal** [comment]
Is this ready for re-review?

**Comment by abdelaziz-mahdy** [comment]
> Is this ready for re-review?

Yes, all the checks and changes are fixed.

File: packages/video_player/video_player/CHANGELOG.md [hunk 7]
```diff
@@ -1,6 +1,7 @@
-## NEXT
+## 2.9.3
 
 * Updates minimum supported SDK version to Flutter 3.22/Dart 3.4.
+* Optimizes caption retrieval with Binary search.
```
Comment (stuartmorgan) [inline packages/video_player/video_player/CHANGELOG.md:4]: binary

File: packages/video_player/video_player/lib/video_player.dart [hunk 8]
```diff
@@ -763,6 +792,14 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
     Future<ClosedCaptionFile>? closedCaptionFile,
   ) async {
     _closedCaptionFile = await closedCaptionFile;
+
+    _sortedCaptions = _closedCaptionFile?.captions;
+
+    /// Sort the captions by start time so that we can do a binary search.
```
Comment (stuartmorgan) [inline packages/video_player/video_player/lib/video_player.dart:798]: "so that we can do" -> "to allow"

Using "we" in comments is often actively confusing for readers, because who "we" is isn't usually clear.

File: packages/video_player/video_player/lib/video_player.dart [hunk 9]
```diff
@@ -763,6 +792,14 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
...
+    _sortedCaptions?.sort((Caption a, Caption b) {
```
Comment (stuartmorgan) [inline packages/video_player/video_player/lib/video_player.dart:799]: `_updateClosedCaptionWithFuture` is called frequently. Re-sorting every caption in the entire file every time this method is called is likely going to hurt overall performance relative to the state before this PR.

The sorted captions should be cached, and only re-generated when the file is changed.

File: packages/video_player/video_player/test/video_player_test.dart [hunk 10]
```diff
@@ -754,7 +776,32 @@ void main() {
...
+            VideoPlayerController.networkUrl(
+          _localhostUri,
+          closedCaptionFile: _loadClosedCaption(),
+        );
+
+        await controller.initialize();
+        final List<Caption> captions =
+            (await controller.closedCaptionFile)!.captions.toList();
+
+        // Check that captions are not in sorted order
```
Comment (stuartmorgan) [inline packages/video_player/video_player/test/video_player_test.dart:790]: This needs a period.

File: packages/video_player/video_player/test/video_player_test.dart [hunk 11]
```diff
@@ -754,7 +776,32 @@ void main() {
     });
 
     group('caption', () {
-      test('works when seeking', () async {
+      test('makes sure the input captions are unsorted', () async {
```
Comment (stuartmorgan) [inline packages/video_player/video_player/test/video_player_test.dart:779]: Why is this a unit test? I.e., why is asserting that the visible caption list is not sorted useful? What user behavior would break if this test were failing?

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/video_player/video_player/test/video_player_test.dart (hunk 11 above)
Comment (abdelaziz-mahdy) [reply packages/video_player/video_player/test/video_player_test.dart:779]: the idea is this test makes sure that the input captions were unsorted, and the tests after it gets the right caption

I am just making sure that no one in the future resorting the input captions, and someone else removing the sorting from inside of the video player in that case the binary search will be broken for unsorted inputs.

please correct me if my thought process is wrong.

File: packages/video_player/video_player/lib/video_player.dart (hunk 9 above)
Comment (abdelaziz-mahdy) [reply packages/video_player/video_player/lib/video_player.dart:797]: I assumed _updateClosedCaptionWithFuture gets called only when file changes, and i cant find any other place it was called, but I made a method that that sorts the captions, and made calls to it when the file changes after `_updateClosedCaptionWithFuture`
I hope thats more efficient now.

Comment (stuartmorgan) [reply packages/video_player/video_player/lib/video_player.dart:797]: > I assumed _updateClosedCaptionWithFuture gets called only when file changes, and i cant find any other place it was called

https://github.com/flutter/packages/blob/a7c1159ac3814bdd4669156bed3f7d5692cece4a/packages/video_player/video_player/lib/video_player.dart#L511C44-L511C68

Comment (abdelaziz-mahdy) [reply packages/video_player/video_player/lib/video_player.dart:797]: noted, so my fix is not accurate, will think of other way and comment in this to let you know

Comment (abdelaziz-mahdy) [reply packages/video_player/video_player/lib/video_player.dart:797]: I hope the added check is good enough, I couldnt think of a better way, since we need to make sure it runs on every new file, do you think I should move back the sorting with the new check in _updateClosedCaptionWithFuture or leaving them seperated is cleaner.

File: packages/video_player/video_player/lib/video_player.dart [hunk 12]
```diff
@@ -378,6 +379,9 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
 
   Future<ClosedCaptionFile>? _closedCaptionFileFuture;
   ClosedCaptionFile? _closedCaptionFile;
+  ClosedCaptionFile? _previousClosedCaptionFile;
```
Comment (stuartmorgan) [inline packages/video_player/video_player/lib/video_player.dart:382]: This is unnecessarily duplicating the memory requirements any time the file changes, and also does an extremely expensive compare for no reason every time. We know when the file is reset, without a compare, because there is an API to set it. We don't need extra state or checks to determine that.

File: packages/video_player/video_player/lib/video_player.dart [hunk 13]
```diff
@@ -378,6 +379,9 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
...
+  List<Caption>? _sortedCaptions;
```
Comment (stuartmorgan) [inline packages/video_player/video_player/lib/video_player.dart:383]: Looking at the `ClosedCaptionFile` type to check the above, I realized it's storing the entire contents of the file in memory, which means that adding this variable is doubling the memory usage of the captioning system.

Do you have any benchmarks about lookup time in a profile build before and after this change with non-trivial files? Given that this is a memory/speed tradeoff, rather than simply an optimization, we need to understand the actual values involved here to know if this is a net positive.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

**Comment by stuartmorgan-g** [comment]
From triage: @abdelaziz-mahdy Are you still planning to update this PR based on the review comments?

**Comment by abdelaziz-mahdy** [comment]
> From triage: @abdelaziz-mahdy Are you still planning to update this PR based on the review comments?

yes, I was busy, will update it and let you know

**Comment by abdelaziz-mahdy** [comment]
i created a mockup benchmark test https://github.com/abdelaziz-mahdy/packages/commit/880c2d243f6d35f44ecc8eff87a06a650bdf9c8c#diff-2030dff5accbb0e6ccf7df8fbd5fc0374da840a562c9594478f997b164a5921e

the results are great even in the case of big files, if we need to eliminate the not sorted to avoid the duplicate, i think we can do that, what do you think? but that will require us to override it i think?
image

**Comment by stuartmorgan-g** [comment]
> if we need to eliminate the not sorted to avoid the duplicate, i think we can do that, what do you think?

Unfortunately:
- the API of the caption file is explicitly that the captions are in file order, and
- the video player has public API to get back the caption file that was set.

I don't see a compelling reason for either of those things to be true, but I'd rather not make a breaking change just for this.

The speedup is definitely worthwhile given those benchmarks, as the lookup in a large file could be a non-trivial portion of a frame allowance. I think the best option here is to track fixing the API in the future (https://github.com/flutter/flutter/issues/168823), but for now duplicate the list for fast lookup.

(Per my earlier comment though, we don't need *two* extra copies of the file, or an expensive compare step to determine when to create a new sorted list. We just need to reset the cached sorted version in the setter.)

**Comment by abdelaziz-mahdy** [comment]
I switched the logic to use `whenComplete` to only call the sort once on the future completion, I hope that matches what you meant, if not I currently don't know how to solve the event listener call to avoid resorting on every event fired,


the plan in my mind currently to add a flag so I know if the current file was sorted or not, but don't know if its good or not.

**Comment by nateshmbhat** [comment]
this will be really cool to have in the subtitle selection @stuartmorgan-g

File: packages/video_player/video_player/lib/video_player.dart [hunk 14]
```diff
@@ -726,20 +729,46 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
   ///
   /// If no [closedCaptionFile] was specified, this will always return an empty
   /// [Caption].
+
   Caption _getCaptionAt(Duration position) {
-    if (_closedCaptionFile == null) {
+    if (_closedCaptionFile == null || _sortedCaptions == null) {
```
Comment (stuartmorgan-g) [inline packages/video_player/video_player/lib/video_player.dart]: Please add a `final List<Captions>? sortedCaptions = _sortedCaptions;` above this, and then use `sortedCaptions` everywhere, so that repeated force-unwraps aren't needed below.

File: packages/video_player/video_player/lib/video_player.dart [hunk 15]
```diff
@@ -726,20 +729,46 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
...
+      },
+    );
+
+    // -1 means not found by the binary search.
+    if (captionIndex == -1) {
+      return Caption.none;
+    }
+
+    final Caption caption = _sortedCaptions![captionIndex];
+    // check if it really fits within that caption's [start, end].
```
Comment (stuartmorgan-g) [inline packages/video_player/video_player/lib/video_player.dart]: Under what circumstances will the `binarySearch` return a result that fails this check? If this is necessary, it seems like the `compare` function above is incorrect.

File: packages/video_player/video_player/lib/video_player.dart [hunk 16]
```diff
@@ -760,8 +789,20 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
   Future<void> _updateClosedCaptionWithFuture(
     Future<ClosedCaptionFile>? closedCaptionFile,
   ) async {
-    _closedCaptionFile = await closedCaptionFile;
-    value = value.copyWith(caption: _getCaptionAt(value.position));
+    if (closedCaptionFile != null) {
+      await closedCaptionFile.whenComplete(() async {
```
Comment (stuartmorgan-g) [inline packages/video_player/video_player/lib/video_player.dart]: Why are you using `whenComplete` in an async function, rather than just `await`ing the future and then using the value?

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]

File: packages/video_player/video_player/lib/video_player.dart (hunk 16 above)
Comment (abdelaziz-mahdy) [reply packages/video_player/video_player/lib/video_player.dart]: To avoid calling the sort Everytime the events get triggered.

Comment (stuartmorgan-g) [reply packages/video_player/video_player/lib/video_player.dart]: I don't follow; every time you call `whenComplete`, the result will be this code running one time.

The way to avoid re-running sorting is to only do the sorting when the file actually changes. As I mentioned in a previous review, we know what that happens, because we controller the code of the setter (`setClosedCaptionFile`).

**Comment by stuartmorgan-g** [comment]
@abdelaziz-mahdy Are you still planning on updating this PR to address the review comments above?

**Comment by abdelaziz-mahdy** [comment]
> @abdelaziz-mahdy Are you still planning on updating this PR to address the review comments above?

Sadly I have been quite busy lately, and not easy to figure out, so I cant fix it yet

**Comment by stuartmorgan-g** [comment]
Marking as a draft pending updates.

**Comment by Piinks** [comment]
Greetings from stale PR triage! 👋 
Is this change still on your radar?

**Comment by abdelaziz-mahdy** [comment]
Review comments addressed. Ready for re-review.

  Changes Made:

  1. Optimized caption lookup

  - Added local variable in _getCaptionAt to avoid repeated force-unwraps
  - Removed unnecessary validity check after binary search (compare function guarantees correctness)

  2. Fixed caching strategy

  - Refactored setClosedCaptionFile to always sort when explicitly setting a new file
  - Updated _updateClosedCaptionWithFuture to only sort on first initialization (_sortedCaptions == null)
  - Prevents expensive re-sorting on every event listener call

  3. Fixed test structure

  - Unnested incorrectly placed tests

  4. Added edge case tests

  Added 5 new tests covering binary search edge cases:
  - Exact caption start/end time boundaries
  - Gaps between captions
  - Single caption scenario
  - Overlapping captions

  The caching optimization ensures captions are only sorted when the file actually changes (initialization or explicit setClosedCaptionFile call), avoiding performance issues.

File: packages/video_player/video_player/lib/video_player.dart [hunk 17]
```diff
@@ -758,20 +761,45 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
...
   /// [Caption].
+
   Caption _getCaptionAt(Duration position) {
-    if (_closedCaptionFile == null) {
+    if (_closedCaptionFile == null || _sortedCaptions == null) {
+      return Caption.none;
+    }
+
+    final List<Caption>? sortedCaptions = _sortedCaptions;
+    if (sortedCaptions == null) {
```
Comment (stuartmorgan-g) [inline packages/video_player/video_player/lib/video_player.dart]: This is now being checked twice; you added a second copy of the check instead of just adding the local before the first copy.

File: packages/video_player/video_player/lib/video_player.dart [hunk 18]
```diff
@@ -785,15 +813,41 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
...
+      _sortedCaptions = List<Caption>.from(_closedCaptionFile!.captions)
+        ..sort((Caption a, Caption b) {
+          return a.start.compareTo(b.start);
+        });
+      value = value.copyWith(caption: _getCaptionAt(value.position));
+    } else {
+      _closedCaptionFile = null;
+      _sortedCaptions = null;
+      value = value.copyWith(caption: Caption.none);
+    }
```
Comment (stuartmorgan-g) [inline packages/video_player/video_player/lib/video_player.dart]: Why have you duplicated all of `_updateClosedCaptionWithFuture` just to change one line?

All you need to do is unconditionally set `sortedCaptions` to null before calling the other method.

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]

File: packages/video_player/video_player/lib/video_player.dart (hunk 18 above)
Comment (abdelaziz-mahdy) [reply packages/video_player/video_player/lib/video_player.dart]: thanks done.

File: packages/video_player/video_player/lib/video_player.dart (hunk 17 above)
Comment (abdelaziz-mahdy) [reply packages/video_player/video_player/lib/video_player.dart]: thanks done.

**Comment by tarrinneal** [comment]
Sorry for the wait, this one slid through my process
